Removing compilation warnings

jvarley opened the pull request:
I started to fix compilation warnings.  Mostly removed unused variables and unitialized variables:

Two alternatives to some of these changes would be:

1)  to add:
-Wno-uninitialized
-Wno-unused

to the QMAKE_CXXFLAGS in graspit-core.pro

2) wrapping them with:
Q_UNUSED

jvarley commented:
@jon-weisz 
@mateiciocarlie 

Let me know what you think is the best way to remove compilation warnings, I will finish the rest of them after we decide what is the best approach.

mateiciocarlie commented on ply/ply.c:
```diff
@@ -1431,7 +1429,7 @@ void ascii_get_element(PlyFile *plyfile, char *elem_ptr)
   int nwords;
   int which_word;
-  char *elem_data,*item;
+  char *elem_data,*item=NULL;
```
These should perhaps use constructors instead of assignments? e.g. char *item(NULL);

mateiciocarlie commented on src/Collision/Graspit/collisionAlgorithms_inl.h:
```diff
@@ -187,7 +187,7 @@ void DistanceCallback::leafTest(const Leaf *l1, const Leaf *l2)
 		for (it2 = list2->begin(); it2!=list2->end() && mMinDistSq >= 0.0; it2++) {
 			mNumTriangleTests++;
-			position p1,p2;
+            position p1=NULL,p2=NULL;
```
I don't think NULL is a legal position. Again, perhaps use a valid constructor (e.g. position p1(0,0,0)). Not entirely sure this is necessary though - does the function call below complain about a possibly uninitialized argument?

jvarley replied:
done.

mateiciocarlie commented on src/Collision/collisionInterface.cpp:
```diff
@@ -164,7 +164,7 @@ CollisionInterface::replaceContactSetWithPerimeter(ContactReport &contactSet)
 	coordT *array = new coordT[contactSet.size()*2];
 	coordT *ptr = &array[0];
-	int ptCount = 0;
+    volatile int ptCount = 0;
```
Why is this needed?

jvarley replied:
from SO:
http://stackoverflow.com/questions/7996825/why-volatile-works-for-setjmp-longjmp

setjmp and longjmp clobber registers. If a variable is stored in a register, its value gets lost after a longjmp.

Conversely, if it's declared as volatile, then every time it gets written to, it gets stored back to memory, and every time it gets read from, it gets read back from memory every time. This hurts performance, because the compiler has to do more memory accesses instead of using a register, but it makes the variable's usage safe in the face of longjmping.

Another alternative solution from same SO article:
Well you could add some extra code to only make it volatile across the call to setjmp. Something like: int x; /* do stuff */ volatile int save_x = x; if(setjmp(buf)) { x = save_x; /* do stuff */ }. This maximizes performance before and after setjmp by using non-volatile variables, but ensures safety by using a volatile variable across the call.

mateiciocarlie replied:
Wow, I never realized what was going on in there...

mateiciocarlie commented on src/EGPlanner/searchEnergy.cpp:
```diff
@@ -356,6 +356,7 @@ SearchEnergy::potentialQualityEnergy(bool verbose) const
 	}
 
 	double gq = -1 ,gv = -1;
+    Q_UNUSED(gv)
```
I would rather clean this up by removing gv as well as the debug statement that uses it altogether...

jvarley replied:
done.

mateiciocarlie commented on src/Planner/grasp_planner.cpp:
```diff
@@ -585,13 +585,13 @@ grasp_planner::get_planningParameters(int& nr_of_360_deg_steps_in,
 SoPathList 
 grasp_planner::searchPrimitives(GraspableBody* bg)
 {	
+  Q_UNUSED(bg)
```
Perhaps remove argument name from function declaration and definition: grasp_planner::searchPrimitives(GraspableBody*)

jvarley replied:
done.

mateiciocarlie commented on src/contact.cpp:
```diff
@@ -1003,7 +1004,7 @@ SoSeparator* SoftContact::getVisualIndicator()
 	zaxisMat->ambientColor = SbColor(0,0,0);
 
-	radius = Body::CONE_HEIGHT / 5;
+    //radius = Body::CONE_HEIGHT / 5;
```
Remove this line instead of commenting out

jvarley replied:
fixed

mateiciocarlie commented on src/eigenGrasp.cpp:
```diff
@@ -592,7 +592,7 @@ EigenGraspInterface::setMinMax()
 	for (int e = 0; e < eSize; e++)
 	{
-		int mind, maxd;
+        //int mind, maxd;
```
Remove what is not needed instead of commenting out

jvarley replied:
fixed

mateiciocarlie commented on src/gws.cpp:
```diff
@@ -114,7 +114,8 @@ int
 {
-  int i,j,k,validPlanes,numCoords,numInLoop;
+  int i,j,k,numCoords,numInLoop;
+  volatile int validPlanes;
```
Why is this needed?

mateiciocarlie commented on src/humanHand.cpp:
```diff
@@ -705,7 +705,7 @@ PROF_DECLARE(TENDON_UPDATE_INSERTION_FORCES);
 {
   PROF_TIMER_FUNC(TENDON_UPDATE_INSERTION_FORCES);
-  SbVec3f pPrev,pCur,pNext;
+  SbVec3f pPrev=NULL,pCur=NULL,pNext=NULL;
```
Not sure NULL is a valid SbVec3f, use a constructor

jvarley replied:
done.

mateiciocarlie commented on src/ivmgr.cpp:
```diff
@@ -652,7 +652,7 @@ IVmgr::transRot(DraggerInfo *dInfo)
 	// save the desired position or orientation and set what percent of the 
 	// desired moved will be accomplished in each step
-	SbVec3f temp = dInfo->lastTran;
+    //SbVec3f temp = dInfo->lastTran;
```
remove if not needed

jvarley replied:
fixed

mateiciocarlie commented:
This generally looks good to me, with a few comments above. 
In general:
* I'd rather remove unneeded stuff instead of commenting out. 
* if a function argument is not needed, you can remove its name in definition and declaration
* I would prefer to avoid QUNUSED just so that we don't introduce gratuitous dependencies on QT
* the biggest problem is of variables needed just for debug code, where the debug output is pre-processed out. QUNUSED might be our best solution to that for now.

jon-weisz commented:
The QUNUSED pramga is trivial.

#define Q_UNUSED(x) (void)x;
We can just add
#define UNUSED(x) (void)x;
someplace.
Not sure where it should be defined though. Maybe in debug.h?

--Jon

>

jvarley commented:
I removed all Q_UNUSED uses except in lmiOptimizer.cpp where it seems appropriate.  I also fixed / commented on all other comments related to this commit.  I believe it is ready to be pulled in.  Let me know if there are any other concerns.

mateiciocarlie commented:
Hmm... I wonder why these latest changes are not showing up in the "Files Changed" section of the Pull Request. Is there a way to update that?

This also raises a much bigger questions that we need to address at this point: unit tests. We would desperately need some, especially for such far reaching changes.

In the past, I informally used two "unit tests":
* running the primitive planner with the Barrett on the mug and confirming that it runs in the expected time, finds the expected top grasp, with the expected quality
* running the dynamics engine on the Barrett on the Puma grasping the glass off the table. Dynamics engine should run, glass should settle in the hand. Then, set glass material to frictionless; glass should get squeezed out of hand.

Since this is a deep change, could you please confirm that these two run as expected?

And we should get to automated unit tests soon...

jvarley commented:
Most recent changes do show up in PR, they are just very small, fixing indentation issues caused by my own early commits, they had looked good in my ide, but indents were off. 
 
I completely agree we need a test framework. Any thoughts on libs to use for it?  I was also thinking of setting up a travis.yml

I ran Menu->Grasp-> Planner with the Barrett and Mug on both the code base with and without this PR and it gets the same results.  Actually the first time I ever ran that Planner, its pretty cool. 

I don't understand how to run your other unit test with/or without this PR.

mateiciocarlie commented:
Second test is as follows:
* load the barrettGlassDyn.xml world
* start dynamics 
* click Autograsp. Glass should be picked up and settle in hand.
* select glass and change material to "frictionless"
* glass should get squeezed out of hand and roll off table. If gravity is on, it falls to the floor.

If you can confirm that this works as expected, I'll go ahead and merge.

Travis is definitely a good option for CI...

jvarley commented:
It works the same for this test both with/without the PR.  One note though for both with/without the PR whenever I turn on dynamics with that world file loaded it spews:

Coin warning in SbRotation::setValue(): rotateTo argument has zero length.
Coin warning in SbRotation::setValue(): rotateTo argument has zero length.
Coin warning in SbRotation::setValue(): rotateTo argument has zero length.
......
Nonstop

But this happens without the PR as well.
